lib/rspec/matchers/expecteds_for_multiple_diffs.rb

```diff
@@ -0,0 +1,118 @@
+module RSpec
+  module Matchers
+    # @api private
+    # Handles list of expected values when there is a need to render
+    # multiple diffs. Also can handle one value.
+    class ExpectedsForMultipleDiffs
+      # @api private
+      # Wraps provided expected value in instance of
+      # ExpectedForMultipleDiffs. If provided value is already an
+      # ExpectedForMultipleDiffs then it just returns it.
+      # @param [Any] expected value to be wrapped
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def self.from(expected)
+        return expected if self === expected
+        new.add(expected)
+      end
+
+      # @api private
+      # Appends expected value to itself. Returns itself for possible
+      # chaining.
+      # @param [Any] expected value to append to itself
+      # @param [Hash] options optional keyword arguments, for example for_matcher
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def add(expected, options={})
+        for_matcher = options.fetch(:for_matcher, NoMatcher.new)
+        expected_list.concat(expected_from(expected, ForMatcher.from(for_matcher)))
+        self
+      end
+
+      # @api private
+      # Returns message with diff(s) appended for provided differ
+      # factory and actual value if there are any
+      # @param [String] message original failure message
+      # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+      # @param [Any] actual value
+      # @return [String]
+      def message_with_diff(message, differ_factory, actual)
+        diff = diffs(differ_factory, actual)
+        message = "#{message}\n#{diff}" unless diff.empty?
+        message
+      end
+
+    protected
+
+      # @api private
+      # Returns list of expected values.
+      # @return [Array<RSpec::Matchers::ExpectedsForMultipleDiffs::Expected>]
+      def expected_list
+        @_expected_list ||= []
+      end
+
+    private
+
+      def diffs(differ_factory, actual)
+        expected_list.map do |expected|
+          expected.diffs(differ_factory, actual)
+        end.inject([], :+).join("\n")
+      end
+
+      def expected_from(value, for_matcher)
+        return value.expected_list if self.class === value
+        [Expected.new(value, for_matcher)]
+      end
+
+      Expected = Struct.new(:value, :for_matcher)
+      # @api private
+      # Represents expected value with possible knowledge of original
+      # matcher
+      class Expected
+        # @api private
+        # Returns diffs for provided differ factory and actual value
+        # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+        # @param [Any] actual value
+        # @return [String]
+        def diffs(differ_factory, actual)
+          diff = differ_factory.call.diff(actual, value)
+          return [] if diff.empty?
+          ["#{for_matcher.diff_headline}#{diff}"]
+        end
+      end
+
+      ForMatcher = Struct.new(:matcher)
```

Comment:
`ForMatcher` instances aren't really value object.  Using `Struct` is handy when you want the `==`/`eql?`/`hash` semantics it provides for value objects, but here they are unused, so I'd prefer to just use a simple class def.

All that said...as much as I like defining a polymorphic interface and multiple small implementations, as you've done here...it feels like a lot of code and ceremony for something that could be much simpler.  For example, `ExpectedsForMultipleDiffs` could have two factory methods:

``` ruby
class ExpectedsForMultipleDiffs
  DEFAULT_DIFF_LABEL = "Diff:".freeze

  def self.from(expected)
    return expected if self === expected
    new([expected, DEFAULT_DIFF_LABEL)
  end

  def self.for_many_matchers(matchers)
    new(matchers.map { |m| [m.expected, diff_label_for(m)] })
  end
end
```

A few things to note about this:

* I intend `for_many_matchers` be the factory method used from `CompoundMatcher`.  Notice that this makes the class immutable, and simplifies the calling interface -- rather than passing `expected` and `matcher` separately (and/or the description string), it just passes the matcher, and let's `for_many_matchers` worry about extracting `expected` and the diff label from it.
* No need for the polymorphic `ForMatcher` and `NoMatcher` classes -- the difference is immediately dealt with at the point of instantiation very simply.
* For something as simple as this, I think I prefer keeping the expected value and diff label together as a simple tuple (e.g. an array of 2 values) rather than making the `Expected` class.  The class feels like overkill when it only provides one method, and I suspect that using arrays will probably perform better since ruby has optimized arrays in C so much.  Plus, these sorts of tuples work really elegantly when later iterating over them since you can destructure into the two elements directly in the block args, e.g.:

``` ruby
diffs = expected_list.map do |(expected_value, diff_label)|
  diff_for(expected_value, diff_label)
end.join("\n")
```

Thoughts?